### PR TITLE
Development complete

### PR DESCRIPTION
**Pull request by eth23git**

### CODE REVIEW

**Comment by danielprypliasko** [comment]
do u need another review lol

**Comment by M-Davenport** [comment]
i did one of them last night so the only review i need is yours

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by eth23git** [comment]
> do u need another review lol

yea please if u can, ill do one back

File: CMP1903_A1_2324/Die.cs [hunk 1]
```diff
@@ -8,16 +8,15 @@ namespace CMP1903_A1_2324
...
     internal class Die
     {
-        /*
-         * The Die class should contain one property to hold the current die value,
-         * and one method that rolls the die, returns and integer and takes no parameters.
-         */
-
         //Property
-
+        private int CurrentValue; //Defines CurrentValue variable
```
Comment (M-Davenport) [inline CMP1903_A1_2324/Die.cs:12]: Good use of encapsulation

File: CMP1903_A1_2324/Die.cs [hunk 2]
```diff
@@ -8,16 +8,15 @@ namespace CMP1903_A1_2324
...
+        private static Random value = new Random(); //Creates a random instance that is accessed 3 times as opposed to making 3 random intances to avoid identical generations.
 
         //Method
-
-
+        public int Roll()
+        {
+            CurrentValue = value.Next(1, 7); //Pseudo-random value between 1 (inclusive) and 7 (exclusive) aka 1-6.
```
Comment (M-Davenport) [inline CMP1903_A1_2324/Die.cs:18]: To make the code more robust it might be useful to do some error handling here in case the value.Next(1, 7) fails.

**Review by danielprypliasko — COMMENTED** [review]
The code meets the requirements of the assignment, 3 Die objects are made, they are rolled and the code is tested well.
However, it seems some variables do not meet the style guide, such as private variables not being prefixed with an underscore nor being camelCase. XML comments seem to also be missing.
The code is well structured and very easy to read, different errors are handled correctly by the Testing class and this makes an efficient program.

**Comment by danielprypliasko** [comment]
> > do u need another review lol
> 
> yea please if u can, ill do one back

added u as a reviewer and left a review

**Comment by M-Davenport** [comment]
cheers. i’m doing yours rn

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: CMP1903_A1_2324/Testing.cs [hunk 3]
```diff
@@ -8,13 +9,20 @@ namespace CMP1903_A1_2324
...
+        //Properties
+        private int total;
 
-        //Method
+        public Testing() //Constructor method to test other classes
+        {
+            Die die = new Die();
+            int roll = die.Roll();
+            Debug.Assert(roll < 7, "Roll value > 6."); //Ensures die roll is within 1-6.
+            Debug.Assert(roll > 0, "Roll value < 0.");
```
Comment (M-Davenport) [inline CMP1903_A1_2324/Testing.cs:20]: good use of error handling to make sure the dice rolls are between 1-6.

File: CMP1903_A1_2324/Die.cs [hunk 4]
```diff
@@ -8,16 +8,15 @@ namespace CMP1903_A1_2324
...
     {
-        /*
-         * The Die class should contain one property to hold the current die value,
-         * and one method that rolls the die, returns and integer and takes no parameters.
-         */
-
         //Property
-
+        private int CurrentValue; //Defines CurrentValue variable
+        private static Random value = new Random(); //Creates a random instance that is accessed 3 times as opposed to making 3 random intances to avoid identical generations.
```
Comment (M-Davenport) [inline CMP1903_A1_2324/Die.cs:13]: Very descriptive comment

File: CMP1903_A1_2324/Program.cs [hunk 5]
```diff
@@ -8,12 +8,14 @@ namespace CMP1903_A1_2324
...
-             * Create a Game object and call its methods.
-             * Create a Testing object to verify the output and operation of the other classes.
-             */
+            Game game = new Game(); //Instntiates new game object
+            game.RollDie(); //Calls method to roll die
+            Console.ReadKey(); //Keeps output window from closing until keystroke.
+
+            Testing test = new Testing(); //Creates testing object where tests run in constructor upon instantiation.
+            Console.ReadKey(); 
         }
```
Comment (M-Davenport) [inline CMP1903_A1_2324/Program.cs:19]: Excellent use of comments on this block of code which greatly improves the readability of the code.

File: CMP1903_A1_2324/Program.cs [hunk 6]
```diff
@@ -8,12 +8,14 @@ namespace CMP1903_A1_2324
 {
     internal class Program
     {
-        static void Main(string[] args)
+        static void Main(string[] args) //Main program body
```
Comment (M-Davenport) [inline CMP1903_A1_2324/Program.cs:11]: This method is clear and concise making the code easy to read and understand.

**Comment by danielprypliasko** [comment]
> > do u need another review lol
> 
> yea please if u can, ill do one back

wait um whats ur student id?

File: CMP1903_A1_2324/Game.cs [hunk 7]
```diff
@@ -1,21 +1,42 @@
...
+        public int roll2;
+
+        private Die die3;
+        public int roll3;
+
+        public int total;
+            
+
+        // Method
+        public void RollDie()
```
Comment (M-Davenport) [inline CMP1903_A1_2324/Game.cs:27]: Clear name for the method making it easy to understand what the method does

File: CMP1903_A1_2324/Game.cs [hunk 8]
```diff
@@ -1,21 +1,42 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Runtime.CompilerServices;
 using System.Text;
 using System.Threading.Tasks;
 
 namespace CMP1903_A1_2324
 {
     internal class Game
```
Comment (M-Davenport) [inline CMP1903_A1_2324/Game.cs:10]: Good use of encapsulation

File: CMP1903_A1_2324/Game.cs [hunk 9]
```diff
@@ -1,21 +1,42 @@
...
+        private Die die3;
+        public int roll3;
+
+        public int total;
+            
+
+        // Method
+        public void RollDie()
+        { 
+            die1 = new Die(); //Instantiates die objects
```
Comment (M-Davenport) [inline CMP1903_A1_2324/Game.cs:29]: good use of instantiation.

File: CMP1903_A1_2324/Game.cs [hunk 10]
```diff
@@ -1,21 +1,42 @@
...
+            roll2 = die2.Roll();
+
+            die3 = new Die();
+            roll3 = die3.Roll();
+
+            total = roll1 + roll2 + roll3;
+            Console.WriteLine("Total roll = " + total); 
+        }
     }
 }
```
Comment (M-Davenport) [inline CMP1903_A1_2324/Game.cs:42]: For this block of code you could add more comments and add more detail to already existing comments.

File: CMP1903_A1_2324/Game.cs [hunk 11]
```diff
@@ -1,21 +1,42 @@
...
-         * The Game class should create three die objects, roll them, sum and report the total of the three dice rolls.
-         *
-         * EXTRA: For extra requirements (these aren't required though), the dice rolls could be managed so that the
-         * rolls could be continous, and the totals and other statistics could be summarised for example.
-         */
 
-        //Methods
+        // Properties
+        private Die die1;
+        public int roll1;
```
Comment (M-Davenport) [inline CMP1903_A1_2324/Game.cs:15]: Good use of suitable variable names.

File: CMP1903_A1_2324/Testing.cs [hunk 12]
```diff
@@ -8,13 +9,20 @@ namespace CMP1903_A1_2324
...
-         */
+        //Properties
+        private int total;
 
-        //Method
+        public Testing() //Constructor method to test other classes
+        {
+            Die die = new Die();
+            int roll = die.Roll();
+            Debug.Assert(roll < 7, "Roll value > 6."); //Ensures die roll is within 1-6.
```
Comment (M-Davenport) [inline CMP1903_A1_2324/Testing.cs:19]: Could add a more descriptive error message

**Review by M-Davenport — COMMENTED** [review]
The code is easy to read but you could improve this further by adding more comments and making already existing comments more descriptive. You have used error handling but it could be made better by making the error messages more descriptive and expanding the testing method to test for different scenarios. You have met all the requirements. You have also generally followed the style guidelines. OOP principles have been used well in this code in particular encapsulation and object instantiation have been effectively implemented.

**Comment by eth23git** [comment]
> > > do u need another review lol
> > 
> > 
> > yea please if u can, ill do one back
> 
> added u as a reviewer and left a review

done urs too now, whats your student id and name?? im 28004501 Ethan Stephenson

**Comment by danielprypliasko** [comment]
> > > > do u need another review lol
> > > 
> > > 
> > > yea please if u can, ill do one back
> > 
> > 
> > added u as a reviewer and left a review
> 
> done urs too now, whats your student id and name?? im 28004501 Ethan Stephenson

27874240 Daniel Prypliasko

**Pull request merged**